---
title: "Basic Bioinformatics Report"
date: \today
output: 
    pdf_document:
        keep_tex: false
        toc: true
        toc_depth: 4
---

```{r setup, echo = FALSE, message = FALSE, warning = FALSE}
knitr::opts_chunk$set(
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  cache = FALSE,
  fig.align = "center"
  )
library(pheatmap)
library(pander)
library(qiimer)
library(vegan)
library(ape)
library(tidyverse)
library(colorRamps)
library(RColorBrewer)
library(ggrepel)
library(viridis)
```

```{r filepaths}
# mapping_FP <- file.path(snakemake@params[["metadata"]])
# preprocess_FP <- file.path(snakemake@params[["preprocess"]])
# fastqc_quality_FP <- file.path(snakemake@params[["fastq_quality"]])
# otu_FP <- file.path(snakemake@params[["kraken"]])

mapping_FP <- "/home/leej39/basic_bioinfo_report/metadata/tobacco_oral_shallow_metadata.tsv"
preprocess_FP <- "/home/leej39/basic_bioinfo_report/metadata/preprocess_summary.tsv"
fastqc_quality_FP <- "/home/leej39/basic_bioinfo_report/metadata/fastqc_quality.tsv"
otu_FP <- "/home/leej39/basic_bioinfo_report/metadata/all_samples.tsv"
```

```{r constants}
richness_subsample_size <- 10000
min_reads <- 100000
```

```{r read_data}
# mapping file
s <- read_tsv(mapping_FP) 

# preprocess 
pre <- read_tsv(preprocess_FP) %>%
  rename(SampleID = Samples) %>%
  filter(SampleID %in% s$SampleID)

# fastqc quality
fq <- read_tsv(fastqc_quality_FP) %>%
  rename(filename = Samples) %>%
  mutate(SampleID = gsub("_R1$|_R2$", "", filename)) %>%
  filter(SampleID %in% s$SampleID)

# OTU
o <- read_qiime_otu_table(otu_FP)

# Taxonomy
a <- o$metadata
a <- sub("(; [kpcofgs]__)+$", "", a, perl = T)
adf0 <- split_assignments(a)
adf <- adf0 %>% unite(Genus_Species, Genus, Species, sep = " ", remove = F) %>% 
  rownames_to_column(var = "taxa_id") %>%
  mutate(Genus_Species = gsub("NA NA", NA, Genus_Species)) %>%
  mutate(Genus_Species = gsub(" NA", "", Genus_Species)) 
rownames(adf) <- adf$taxa_id
adf <- adf %>% 
  select(Kingdom, Phylum, Class, Order, Family, Genus_Species, Genus, Species)

# OTU table
counts <- o$counts
counts_col <- colnames(counts)
counts_col <- gsub(".taxa", "", counts_col, fixed = T)
colnames(counts) <- counts_col
counts <- counts[rownames(adf), s$SampleID]

# Bacteria
is.bacteria <- grepl("k__Bacteria", adf$Kingdom)
counts_bacteria <- counts[is.bacteria,]
props_bacteria <- sweep(counts_bacteria, 2, colSums(counts_bacteria), "/")
adf_bacteria <- adf[is.bacteria,]
a_bacteria <- simplify_assignments(adf_bacteria, rank1 = "Phylum", rank2 = "Genus_Species")

bacteria_Read_Counts <- colSums(counts_bacteria) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  setNames(c("SampleID", "Bacteria_Read_Counts"))

# read counts after demuxing/both kept/non_host/bacteria
s <- s %>% left_join(pre[, c("SampleID", "input", "both_kept", "false", "true")], by = "SampleID") %>%
  rename(demux_Read_Counts = input) %>%
  rename(both_kept_Read_Counts = both_kept) %>%
  rename(non_host_Read_Counts = false) %>%
  left_join(bacteria_Read_Counts, by = "SampleID") 

# alpha diversity
s <- s %>% 
  merge(diversity(t(counts_bacteria)), by.x = "SampleID", by.y = "row.names", all.x = T) %>%
  rename(shannon = y) %>%
  merge(rarefy(t(counts_bacteria), richness_subsample_size), by.x = "SampleID", by.y = "row.names", all.x = T) %>%
  rename(richness = y) 

# additional data prep
s <- s %>%
  mutate(SubjectID = factor(SubjectID))
```

\newpage

# Average nucleotide quality after adapter trimming and quality control

Average nucleotide quality score after trimming adapter and filtering low quality (only forward read or reverse read or neither was kept after trimmomatic) read pairs is given. Red dots represent the mean quality score and vertical line segments represent mean $\pm$ sample standard deviation.

```{r, fig.width = 8, fig.height = 4}
rexp <- "^(.*)_(R[12])$"
quality <- fq %>% 
  gather(Position, Quality, -filename, -SampleID) %>%
  mutate(Position = sub("-\\d+", "", Position, perl = TRUE)) %>%
  mutate(Position = as.numeric(Position)) %>%
  mutate(Direction = sub(rexp,"\\2", filename)) %>%
  mutate(Direction = factor(Direction))

ave_quality <- quality %>%
  group_by(Direction, Position) %>%
  summarise(MeanQual = mean(Quality), SdQual = sd(Quality)) %>%
  mutate(LowQual = MeanQual - SdQual, HighQual = MeanQual + SdQual)

ggplot(ave_quality, aes(Position, MeanQual)) + 
  geom_errorbar(aes(ymin = LowQual, ymax = HighQual)) +
  theme(aspect.ratio = 1) + 
  facet_wrap(~ Direction) +
  geom_point(color = "red") +
  labs(x = "Position in sequence read", y = "Average quality score")
```

\newpage

# Read Counts

Red dashed lines represent 1 million counts. 

```{r read_counts, fig.height = 7}
g <- s %>%
  gather(key = "CountType", value = "count", ends_with("Read_Counts")) %>%
  mutate(CountType = factor(CountType, 
                            levels = c("demux_Read_Counts",
                                       "both_kept_Read_Counts",
                                       "non_host_Read_Counts",
                                       "Bacteria_Read_Counts"))) %>%
  ggplot(aes(study_group, count)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_hline(yintercept = 1e6, lty = "dashed", color = "red") +
  geom_jitter() +
  facet_wrap(~CountType, scales = "free_y")

print(g)
```

\newpage

# Heatmap

```{r}
prop_cut <- 0.01
satu_limit <- 0.4
```

Each column of the heatmap represents one sample and each row represents one taxon. Taxa were included in the chart if the abundance in any sample exceeded `r 100*prop_cut`%. 

The chart is colored white if taxa were not observed in the sample, dark blue if taxa were observed at very low abundance. This allows the reader to quickly survey presence/absence. Abundance values exceeding `r 100*satu_limit`% are colored red, indicating an extremely dominant species.

```{r heatmap, fig.width = 20, fig.height = 24}
s_Heat <- s %>%
  as.data.frame() # "annotation" of pheatmap() should be a dataframe

grps <- c("study_group")
s_Heat <- s_Heat[, c("SampleID", grps)] 
s_Heat <- arrange_(s_Heat, .dots = grps)

anno <- s_Heat %>% 
  select(-SampleID) 
rownames(anno) <- s_Heat$SampleID
colnames(anno) <- grps

GS_counts <- rowsum(counts_bacteria, a_bacteria)
GS_props <- sweep(GS_counts, 2, colSums(GS_counts), "/")
props_Heat <- GS_props[, s_Heat$SampleID]
taxa_names_to_Heat <- rownames(props_Heat)[apply(props_Heat, 1, max) >= prop_cut]
props_Heat <- props_Heat[taxa_names_to_Heat, ]

color = saturated_rainbow(101, saturation_limit = satu_limit)
breaks = c(0, 1e-10, seq(0.001, 1, length.out = 100))

pheatmap(props_Heat, annotation = anno, cluster_cols = F, cluster_rows = F, 
         color = color, breaks = breaks,  
         cellwidth = 8, cellheight = 8, fontsize_col = 8, fontsize_row = 8)
```

\newpage

# Alpha diversity

Alpha diversity (within sample diversity) was assessd by richness and Shannon index.

```{r}
s_Alpha <- s %>%
  gather(key = "alpha_measure", value = "alpha_value", richness, shannon)
```

```{r, fig.height = 4}
g <- ggplot(s_Alpha, aes(x = study_group, y = alpha_value)) +
  geom_boxplot(outlier.alpha = 0) +
  ylim(0, NA) +
  theme(aspect.ratio = 1, axis.text.x = element_text(angle = -60, hjust = 0, vjust = 0.5)) +
  facet_wrap(~alpha_measure, ncol = 2, scales = "free_y")

print(g)
```

\newpage

# Beta diversity

Beta diversity (similarity between samples) was assessed by Bray-Curtis distance and Jaccard distance. 

```{r}
dist_in <- vegdist(t(props_bacteria))
dist_name <- "Bray-Curtis distance"
```

## Ordination based on `r dist_name`

```{r, fig.height = 3.5}
s_Beta <- s 
showup_both <- intersect(colnames(as.matrix(dist_in)), s_Beta$SampleID)
dist_in <- dist_subset(dist_in, showup_both)
s_Beta <- s_Beta[match(showup_both, s_Beta$SampleID),]

pcoa_title <- paste0(dist_name)
pcoa_result <- pcoa(dist_in)
pcoa_df <- merge(s_Beta, pcoa_result$vectors[,1:2], by.x = "SampleID", by.y = "row.names")
pcoa_pct <- round(pcoa_result$values$Relative_eig*100, 1)

pcoa_plot <- ggplot(pcoa_df, aes(x = Axis.1, y = Axis.2)) +
  theme_bw() +
  xlab(paste0("PCoA axis 1 (", pcoa_pct[1], "%)")) +
  ylab(paste0("PCoA axis 2 (", pcoa_pct[2], "%)")) +
  ggtitle(pcoa_title) + 
  geom_point(aes(color = SubjectID, shape = study_day)) +
  theme(aspect.ratio=1) 

print(pcoa_plot)
```

```{r}
dist_in <- dist(t(props_bacteria), method = "binary")
dist_name <- "Jaccard distance"
```

## Ordination based on `r dist_name`

```{r, fig.height = 3.5}
s_Beta <- s 
showup_both <- intersect(colnames(as.matrix(dist_in)), s_Beta$SampleID)
dist_in <- dist_subset(dist_in, showup_both)
s_Beta <- s_Beta[match(showup_both, s_Beta$SampleID),]

pcoa_title <- paste0(dist_name)
pcoa_result <- pcoa(dist_in)
pcoa_df <- merge(s_Beta, pcoa_result$vectors[,1:2], by.x = "SampleID", by.y = "row.names")
pcoa_pct <- round(pcoa_result$values$Relative_eig*100, 1)

pcoa_plot <- ggplot(pcoa_df, aes(x = Axis.1, y = Axis.2)) +
  theme_bw() +
  xlab(paste0("PCoA axis 1 (", pcoa_pct[1], "%)")) +
  ylab(paste0("PCoA axis 2 (", pcoa_pct[2], "%)")) +
  ggtitle(pcoa_title) + 
  geom_point(aes(color = SubjectID, shape = study_day)) +
  theme(aspect.ratio=1) 

print(pcoa_plot)
```

\newpage

# Appendix

## Number of reads before and after trimmming Illumina adapter sequences with Trimmomatic.

```{r}
temp <- pre %>%
  arrange(both_kept) 

pander(temp, caption = "number of reads before and after Trimmomatic")
```

\newpage

## Number of reads before and after filtering of host genome sequence.

```{r}
temp2 <- pre %>%
  mutate(true = replace(true, is.na(true), 0)) %>%
  mutate(
    `Percent host reads` = 100 * true / (true + false),
    `Percent host reads` = round(`Percent host reads`, 2)) %>%
  select(
    SampleID,
    `Host reads` = true,
    `Non-host reads` = false,
    `Percent host reads`) %>%
  arrange(`Non-host reads`)

pander(temp2, caption = "Number of reads before and after removing host sequences")
```


